dev/OAuth/OAuth2Manager.cpp

```diff
@@ -0,0 +1,458 @@
+                {
+                    for (auto&& pair : map)
+                    {
+                        if (!headers.TryAppendWithoutValidation(pair.Key(), pair.Value()))
+                        {
+                            // TODO? Why might this fail? Throw?
+                        }
+                    }
+                }
+            }
+
+            auto lifetime{ get_strong() };
+            auto cancellation = co_await winrt::get_cancellation_token();
+            cancellation.enable_propagation();
+
+            response = co_await httpClient.SendRequestAsync(request);
+            // TODO: Check status code?
+            if (!response.IsSuccessStatusCode())
+            {
+                auto status = response.StatusCode();
+                HRESULT hr = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_HTTP, static_cast<HRESULT>(status));
+                co_return implementation::TokenRequestResult::MakeFailure(std::move(response),
+                    TokenFailureKind::HttpFailure, hr);
+            }
+
+            auto responseContentType = response.Content().Headers().ContentType().MediaType();
+            if (responseContentType != L"application/json")
+            {
+                co_return implementation::TokenRequestResult::MakeFailure(std::move(response),
+                    TokenFailureKind::InvalidResponse, WEB_E_UNSUPPORTED_FORMAT);
+            }
+
+            responseString = co_await response.Content().ReadAsStringAsync();
+        }
+        catch (...)
+        {
+            co_return implementation::TokenRequestResult::MakeFailure(std::move(response),
+                TokenFailureKind::HttpFailure, winrt::to_hresult());
+        }
+
+        JsonObject jsonObject{ nullptr };
+        if (!JsonObject::TryParse(responseString, jsonObject))
+        {
+            co_return implementation::TokenRequestResult::MakeFailure(std::move(response),
+                TokenFailureKind::InvalidResponse, WEB_E_INVALID_JSON_STRING);
+        }
+        else
+        {
+            try
+            {
+                // Determine if it's a success or error response based on the presence of 'error'
+                if (jsonObject.HasKey(L"error"))
+                {
+                    auto failure = winrt::make<implementation::TokenFailure>(jsonObject);
+                    co_return winrt::make<implementation::TokenRequestResult>(std::move(response), nullptr,
+                        std::move(failure));
+                }
+                else
+                {
+                    auto success = winrt::make<implementation::TokenResponse>(jsonObject);
+                    co_return winrt::make<implementation::TokenRequestResult>(std::move(response), std::move(success),
+                        nullptr);
+                }
+            }
+            catch (...)
+            {
+                co_return implementation::TokenRequestResult::MakeFailure(std::move(response),
```

Comment:
For these "catch and return" points, consider using `LOG_CAUGHT_EXCEPTION()` so you can see what's happening.

Reply:
The official Winappsdk docs mention how we can pull out the relevant hr code and message.
https://github.com/microsoft/WindowsAppSDK/blob/743f4551dcd8eda8db1136d7f61a8ecc7f338325/docs/Coding-Guidelines/Languages-CPP.md

Would it also be helpful if we can fire a telemetry with the error codes for debugging and tracking metrics?

Reply:
Used LOG_CAUGHT_EXCEPTION();